src/components/tl/ucp/allgather/allgather_ring.c

```diff
@@ -0,0 +1,79 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2021.  ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#include "config.h"
+#include "tl_ucp.h"
+#include "allgather.h"
+#include "core/ucc_progress_queue.h"
+#include "tl_ucp_sendrecv.h"
+#include "utils/ucc_math.h"
+#include "core/ucc_mc.h"
+
+ucc_status_t ucc_tl_ucp_allgather_ring_progress(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_ucp_task_t *task       = ucc_derived_of(coll_task, ucc_tl_ucp_task_t);
+    ucc_tl_ucp_team_t *team       = task->team;
+    ucc_rank_t         group_rank = team->rank;
+    ucc_rank_t         group_size = team->size;
+    void              *rbuf       = task->args.dst.info.buffer;
+    ucc_memory_type_t  rmem       = task->args.dst.info.mem_type;
+    size_t             count      = task->args.dst.info.count;
+    ucc_datatype_t     dt         = task->args.dst.info.datatype;
+    size_t             data_size  = count * ucc_dt_size(dt);
+    ucc_rank_t         sendto     = (group_rank + 1) % group_size;
+    ucc_rank_t         recvfrom   = (group_rank - 1 + group_size) % group_size;
+    int                step;
+    if (UCC_INPROGRESS == ucc_tl_ucp_test(task)) {
+        return task->super.super.status;
+    }
+
+    while (task->send_posted < group_size - 1) {
+        step = task->send_posted;
+        ucc_tl_ucp_send_nb((void*)((ptrdiff_t)rbuf +
+                           ((group_rank - step + group_size) % group_size)
```

Comment:
Is it required to send to self for correctness?

Reply:
dest buffer is different from source in general - in that case we need to send_self (we could use mc_memcpy, but adding another branch and reducing code readability - ucp will do memcpy underhoods anyways).